### PR TITLE
Lesson 5

### PR DESCRIPTION
**Pull request by lunarnuts**

### CODE REVIEW

File: lesson05/shapes/circle.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"fmt"
+	"math"
+)
+
+type Circle struct {
+	radius float64
+}
+
+func (c Circle) Area() (float64, bool) {
```
Comment (gordiyanovp) [inline lesson05/shapes/circle.go]: This is not Go idiomatic. The second return value should be error, which needs to be returned on incorrect input (negative radius value in this case). If the error is not nil, the float64 value needs to be default. Same with other methods on Circle and Rectangle

File: lesson05/shapes/circle_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		radius float64
+	}
+	type args struct {
+		r float64
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		want   bool
```
Comment (gordiyanovp) [inline lesson05/shapes/circle_test.go]: If the return value is changed according to my comments above, the want value probably needs to be of type error.

File: lesson05/shapes/circle.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (c Circle) String() string {
+	if c.radius <= 0 {
+		return fmt.Sprintf("radius needs to be positive and non-zero, got: %.3f", c.radius)
+	}
+	return fmt.Sprintf("\nCircle: radius %.2f", c.radius)
+}
+
+func (c Circle) GetRadius() (float64, bool) {
```
Comment (gordiyanovp) [inline lesson05/shapes/circle.go]: Same as above, the method should return an error

File: lesson05/shapes/circle.go (hunk 1 above)
Comment (lunarnuts) [reply lesson05/shapes/circle.go]: fixed

File: lesson05/shapes/circle_test.go (hunk 2 above)
Comment (lunarnuts) [reply lesson05/shapes/circle_test.go]: fixed

File: lesson05/shapes/circle.go (hunk 3 above)
Comment (lunarnuts) [reply lesson05/shapes/circle.go]: fixed

File: lesson05/shapes/circle_test.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+		{"positive radius", fields{radius: 1}, args{r: 1}, nil},
+		{"zero radius", fields{radius: 0}, args{r: -1}, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", 0.0)},
+		{"negative radius", fields{radius: -1}, args{r: -1}, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", -1.0)},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Circle{
+				radius: tt.fields.radius,
+			}
+			if got := c.SetRadius(tt.args.r); got == nil && tt.want != nil || got != nil && tt.want == nil {
```
Comment (gordiyanovp) [inline lesson05/shapes/circle_test.go:31]: You shouldn't have these complicated if statements in your tests. Also, you should check your result against what you want, not nil:

```go
if got != tt.want{
     ...
}
```

Same goes for other tests below

**Review by gordiyanovp — COMMENTED** [review]
You shouldn't really rmerge your PR until you get an approval. There are still things you need to fix right now.

**Pull request merged**